[RFC] run test on stm32F7 with CONFIG_UART_ASYNC_API and DMA

FRASTM opened the pull request:
This patch is disabling the Cache memory control from the stm32F7xx  when the arm cpu is cortex M7 
It comes after the https://github.com/zephyrproject-rtos/zephyr/pull/35258.

On nucleo_f746zg and nucleo_f767zi : USART 6 is used for testing : DMA 2 channel 7 & 2 and request 5 is selected (usart6 and dma2 are already selected for this board)
UART Tx pin must be connected to UART Rx pin on the board to run the `tests/drivers/uart/uart_async_api`.

--> `tests/drivers/uart/uart_async_api` passed  on nucleo _f746zg and nucleo_f767zi boards

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: dma stm32 testing transfer on the nucleo_f767zi

Adding the dma memory-to-memory transfers on dma2,
tests/driver/dma/chan_blen_transfer and
tests/driver/dma/loop_transfer
This board is disabling the CACHE

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Due to the merge of https://github.com/zephyrproject-rtos/zephyr/pull/35258,  the cache on stm32f7 is controlled at boot, not directly in the stm32f7/soc.c

aurel32 commented:
> Due to the merge of #35258, the cache on stm32f7 is controlled at boot, not directly in the stm32f7/soc.c
> The `soc/arm/st_stm32/stm32f7/Kconfig.series` could also select:
> select INIT_ARCH_HW_AT_BOOT

I think there is a big misunderstanding here. The changes in #35258 haven't change anything for the way the cache is enabled on the stm32f7 is handled. It has just fixed the way the cache clean was wrongly done.

The goal of `INIT_ARCH_HW_AT_BOOT` is to bring the hardware in a fully reproducible state, independently of what happened before Zephyr boots (ROM code, bootloader, wake up from sleep mode, etc.). I do not see the point of selecting it in `soc/arm/st_stm32/stm32f7/Kconfig.series`. However this is questionnable if it should be the default, independently of the SoC being used.

aurel32 commented on soc/arm/st_stm32/stm32f7/soc.c:
```diff
@@ -32,11 +32,6 @@ static int st_stm32f7_init(const struct device *arg)
 
 	key = irq_lock();
 
-	SCB_EnableICache();
```
This is not correct at all. The code in `arch/arm/core/aarch32/cortex_m/scb.c` is run very early in the boot process, way before the STM32F7 specific code is used. Its goal is to bring the hardware in a know, reproducible state.

There is still need to enable it later during the boot process in the STM32F7 specific code.

FRASTM replied:
ok, restored cache enable if needed in the stm32f7/soc.c 
add the config for testing the stm32f746zg nucleo board only

aurel32 replied:
First of all the I-Cache can be kept enabled in that case, it doesn't interfere with DMA.

Secondly, I still do not get why we want to disable the cache if CONFIG_NOCACHE_MEMORY is not enabled. It has a huge performance impact, I would see that the other way around, i.e. if that CONFIG_UART_ASYNC_API selects CONFIG_NOCACHE_MEMORY on Cortex M7 or STM32F7.

Alternatively we might want to always enable CONFIG_NOCACHE_MEMORY on Cortex M7 or STM32F7, or enable it when DMA is enabled.

FRASTM commented:
_connect D0 to D1 on CN10 of the nucleo144 board_

```
*** Booting Zephyr OS build zephyr-v2.5.0-4014-g9974db5a1861  ***               
Running test suite uart_async_test                                              
===================================================================             
START - test_single_read_setup                                                  
 PASS - test_single_read_setup in 0.1 seconds                                   
===================================================================             
START - test_single_read                                                        
 PASS - test_single_read in 0.403 seconds                                       
===================================================================             
START - test_chained_read_setup                                                 
 PASS - test_chained_read_setup in 0.1 seconds                                  
===================================================================             
START - test_chained_read                                                       
 PASS - test_chained_read in 0.269 seconds                                      
===================================================================             
START - test_double_buffer_setup                                                
 PASS - test_double_buffer_setup in 0.1 seconds                                 
===================================================================             
START - test_double_buffer                                                      
 PASS - test_double_buffer in 3.428 seconds                                     
===================================================================             
START - test_read_abort_setup                                                   
 PASS - test_read_abort_setup in 0.1 seconds                                    
===================================================================             
START - test_read_abort                                                         
 PASS - test_read_abort in 1.61 seconds                                         
===================================================================             
START - test_chained_write_setup                                                
 PASS - test_chained_write_setup in 0.1 seconds                                 
===================================================================             
START - test_chained_write                                                      
 PASS - test_chained_write in 0.3 seconds                                       
===================================================================             
START - test_long_buffers_setup                                                 
 PASS - test_long_buffers_setup in 0.1 seconds                                  
===================================================================             
START - test_long_buffers                                                       
 PASS - test_long_buffers in 0.153 seconds                                      
===================================================================             
START - test_write_abort_setup                                                  
 PASS - test_write_abort_setup in 0.1 seconds                                   
===================================================================             
START - test_write_abort                                                        
 PASS - test_write_abort in 0.103 seconds                                       
===================================================================             
START - test_forever_timeout_setup                                              
 PASS - test_forever_timeout_setup in 0.1 seconds                               
===================================================================             
START - test_forever_timeout                                                    
 PASS - test_forever_timeout in 3.2 seconds                                     
===================================================================             
Test suite uart_async_test succeeded                                            
===================================================================             
PROJECT EXECUTION SUCCESSFUL    
```

FRASTM commented:
In the test_uart_async.c , when trying to define tx_buf in Nocache memory area, like the following:
```
static __aligned(16) uint8_t tx_buf[128] __used
	__attribute__((__section__(".nocache")));
```

there is a MPU fault
```
E: ***** MPU FAULT *****
E:   Data Access Violation
E:   MMFAR Address: 0x20020420
E: r0/a1:  0x20020420  r1/a2:  0x08010812  r2/a3:  0x00000004
E: r3/a4:  0x20020420 r12/ip:  0x0800d3ff r14/lr:  0x080009e7
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x0800db00
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20021200 (test_single_read)
E: Halting system
```

MMFAR Address: 0x20020420 is the tx_buf address (in supposed non-cached mem.)

FRASTM committed:
aarch32: core: mpu allow RW access for USER in NOCACHE_MEMORY

The attribut of the Special non-cacheable RAM area is set
to RW for USERSPACE. In this dedicated area, the user application
can access non-cached buffers defined for the peripheral transfer
with DMA.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
With this new version, the rx and tx buffers are allocated in the NON-CACHED memory area when the DMA is enabled for uart transfers.
On the stm32f767 and stm32f746 nucleo boards, when the CONFIG_NOCACHE_MEMORY=y, the testcase is modified like in the tests/drivers/dma/chan_blen_transfer. Consequently the /home/r/zephyrproject/zephyr/soc/arm/st_stm32/stm32f7/soc.c can enable I-cache and D-cache without disturbing the dma transfer.

FRASTM commented:
Commit https://github.com/zephyrproject-rtos/zephyr/pull/32833/commits/a9696c327d57aeb9c73b8ad788086fa3fe076b4c is required to give to the USERSPACER a RW access onf the non-cached memory where dma Rx-Tx buffers are allocated. 
All the NOCACHE_MEMORY attribut is modified.

erwango commented on arch/arm/core/aarch32/mpu/arm_core_mpu.c:
```diff
@@ -76,7 +76,7 @@ static const struct z_arm_mpu_partition static_regions[] = {
 		.start = (uint32_t)&_nocache_ram_start,
 		.size = (uint32_t)&_nocache_ram_size,
-		.attr = K_MEM_PARTITION_P_RW_U_NA_NOCACHE,
+		.attr = K_MEM_PARTITION_P_RW_U_RW_NOCACHE,
```
@ioannisg, @aurel32  Would you mind giving your thoughts on this ?
Change shows that a user space accessible non cacheable partition would make sense for M7 userspace DMA users (simplify use of DMA base transaction while keeping cache related performances, as for now the alternative is to fully disable cache)
So, would it be acceptable to change attributes of this partition ?
If not, would it make sense to add a new partition that would be userspace RW ?

aurel32 replied:
> @ioannisg, @aurel32 Would you mind giving your thoughts on this ?
> Change shows that a user space accessible non cacheable partition would make sense for M7 userspace DMA users (simplify use of DMA base transaction while keeping cache related performances, as for now the alternative is to fully disable cache)
> So, would it be acceptable to change attributes of this partition ?

When we discussed adding the NO_CACHE memory, we had in mind that it's only required at the driver levels, that's why we only make it accessible to kernel space. I don't think we can change that, it basically gives access to memory areas that the user space is not supposed to get access to (network data for example), so this is a security issue.

> If not, would it make sense to add a new partition that would be userspace RW ?

That's an option, to consider.

The other alternative is to use cache management operation at the driver level. This is already implemented in some drivers (Atmel Ethernet driver, Atmel and STM32 I2S drivers), and I guess the same can be done for the serial driver. Compare to uncached memory, that should bring some performance improvement, and simpler userland code, in the sense that the memory only need to be declared aligned, not anymore in the nocache area. It's just a pitty we still do not have functions for that in Zephyr and that we have to rely on the CMSIS functions.

aurel32 replied:
We clearly lack cache related functions. I guess the same way we have created the `__nocache__` macro, we can create something like a `__dma_alloc__` macro that takes care of aligning the data correctly. This should not be different from the user point of view.

Adding a new MPU region for uncached memory is not trivial either, at least in the `CONFIG_USERSPACE=y`, because you need to swap that MPU region on task change. It also mean we need to create a new __nocache_user__ macro to put the variables in the right sections.

The other short term alternatives is to disable the D-Cache (but not the I-Cache) when `CONFIG_UART_ASYNC_API=y`. 

An other STM32 specific short term alternative is to create `__dma_alloc__` macro that allocates the variables in the DTCM. In the longer term we can imagine to let users choose between allocating those variables in the DTCM or in standard memory with the use of cache management operations.

aurel32 replied:
Some additional thinking:
- `__nocache` prevents 1) variable allocated on the stack (the section is silently dropped) 2) initialized. This explains why the uart_async_api test needs to be heavily modified.
- Cache management operations requires that 1) buffer are aligned on a cache line (32-byte) 2) There is proper padding so that no other unrelated variables are placed on the same cache line. I *think* those requirements are actually only true for RX buffer, as a D-Cache clean operation doesn't have side effects besides performance.

Both cases definitely have an impact on the user code, although in a different way.

erwango replied:
> The other short term alternatives is to disable the D-Cache (but not the I-Cache) when CONFIG_UART_ASYNC_API=y.

That's the easiest indeed. But we need to do it also for other DMA users like SPI, I2S and upcoming ones (hopefully ADC).
Eventually, only some happy few will benefit from full M7 cache performances :-).

> An other STM32 specific short term alternative is to create __dma_alloc__ macro that allocates the variables in the DTCM. 

That's indeed an interesting direction as well. Ideally we can extend that to other M7 + DTCM based architectures.

Thanks @aurel32 for your advised feedback. Let's see what @ioannisg think about that.

FRASTM replied:
Defining a contiguous USER NOCACHE memory area to store (`__section__(".usernocache")`) user buffers is possible for this test :
```
                0x0000000020020000                . = 0x20020000
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                _image_ram_start = .

nocache         0x0000000020020000     0x1000
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                . = ALIGN ((0x1 << LOG2CEIL (_nocache_ram_size)))
                0x0000000020020000                _nocache_ram_start = .
 *(SORT_BY_ALIGNMENT(.nocache))
 *(SORT_BY_ALIGNMENT(.nocache.*))
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                . = ALIGN ((0x1 << LOG2CEIL (_nocache_ram_size)))
                0x0000000020020000                _nocache_ram_end = .
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                . = ALIGN ((0x1 << LOG2CEIL (_nocache_ram_size)))
                0x0000000020020000                _usernocache_ram_start = .
 *(SORT_BY_ALIGNMENT(.usernocache))
 .usernocache   0x0000000020020000      0xf10 app/libapp.a(test_uart_async.c.obj)
 *(SORT_BY_ALIGNMENT(.usernocache.*))
                0x0000000020020f20                . = ALIGN (_region_min_align)
 *fill*         0x0000000020020f10       0x10 
                0x0000000020021000                . = ALIGN ((0x1 << LOG2CEIL (_usernocache_ram_size)))
 *fill*         0x0000000020020f20       0xe0 
                0x0000000020021000                _usernocache_ram_end = .
                0x0000000000000000                _nocache_ram_size = (_nocache_ram_end - _nocache_ram_start)
                0x0000000020020000                _usernocache_ram_start = _nocache_ram_end
                0x0000000000001000                _usernocache_ram_size = (_usernocache_ram_end - _usernocache_ram_start)
```

ABOSTM commented on tests/drivers/uart/uart_async_api/boards/nucleo_f746zg.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_NOCACHE_MEMORY=y
```
Would it be interesting to have an error or a warning at kconfig level if application doesn't enable this switch (in case of DMA and cache enabled) ?
Message could explain that DMA buffer should be placed in the no cache memory.

aurel32 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -67,26 +73,44 @@ void test_single_read_setup(void)
+		__attribute__((__section__(".nocache")));
+	static __aligned(16) uint8_t rx_buf[16] __used
+		__attribute__((__section__(".nocache.dma")));
+
```
I wonder why this variable need to be aligned to 16 bytes. For an uncached section, there is no need for specific alignment. I also wonder why the difference in section `".nocache"` vs `".nocache.dma"'. Why not simply using the `__nocache` macro?

FRASTM replied:
I tried building the test with a special "USER NOCACHE MEMORY" area where user buffers are defined for the DMA transfers.  The CONFIG_NOCACHE_MEMORY could be ANDed with CONFIG_USERSPACE  to map this area.

FRASTM replied:
We can see 4 different approaches, where item 1 and 2 are feasible
1. allocate a user space NOCACHE MEMORY 
2. Invalidate/clean Cache memory in _each_ driver with DMA (adapt for each peripheral driver with dma)
3. Invalidate/clean Cache memory in the dma drivers : once for all the pepriphera
4. allocate a user space NOCACHE MEMORY for the dma driver and memcpy in each client buffer

FRASTM replied:
should data be aligned on data aligned on `DCACHE_LINE_SIZE` or on `arch_dcache_line_size_get() `

aurel32 replied:
In the NOCACHE memory case, I don't think you need any alignment.

aurel32 replied:
> We can see 4 different approaches, where item 1 and 2 are feasible
> 
>     1. allocate a user space NOCACHE MEMORY
> 
>     2. Invalidate/clean Cache memory in _each_ driver with DMA (adapt for each peripheral driver with dma, like in #27911)
> 
>     3. Invalidate/clean Cache memory in the dma drivers : once for all the peripheral

If going the cache management operation way, the option 3 would be great to avoid redoing the job for each driver.
 
>     4. allocate a user space NOCACHE MEMORY for the dma driver and memcpy in each client buffer

This option also works with DTCM memory or even with standard memory with cache management operation. Note that this could be made optional if the driver doesn't communicate directly with the user space, but rather with a subsystem and the subsystem buffer can be used directly (example network packets for an Ethernet driver).

FRASTM commented:
I prepared a branch including an aditionnal USER NOCACHE space in the NOCACHE memory area   the https://github.com/FRASTM/zephyr/tree/user_cache
On the other hand I was looking for the stm 32 uart driver, cleaning and invalidating the cache memory (like i2s does) in this branch https://github.com/FRASTM/zephyr/tree/uart_cache_coherency
but I get 
```
START - test_chained_write
E: ***** MPU FAULT *****
E:   Instruction Access Violation
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x000000
E: r3/a4:  0x00000000 r12/ip:  0x0800d27b r14/lr:  0x0800d5
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x01010100
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20021000 (test_chained_write)
E: Halting system
```

aurel32 commented:
> I prepared a branch including an aditionnal USER NOCACHE space in the NOCACHE memory area the https://github.com/FRASTM/zephyr/tree/user_cache
> On the other hand I was looking for the stm 32 uart driver, cleaning and invalidating the cache memory (like i2s does) in this branch https://github.com/FRASTM/zephyr/tree/uart_cache_coherency
> but I get
> 
> ```
> START - test_chained_write
> E: ***** MPU FAULT *****
> E:   Instruction Access Violation
> E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x000000
> E: r3/a4:  0x00000000 r12/ip:  0x0800d27b r14/lr:  0x0800d5
> E:  xpsr:  0x21000000
> E: Faulting instruction address (r15/pc): 0x01010100
> E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
> E: Current thread: 0x20021000 (test_chained_write)
> E: Halting system
> ```

I have looked quickly at your branch. Here are some comments:
- The cache invalidate operation can be done in `async_evt_rx_rdy()` to avoid doing it at 2 different places
- I don't see why you need a cache clean operation in `uart_stm32_dma_tx_cb()`
- With cache operations, you need to have the buffer aligned on DCACHE_LINE_SIZE, ie 32 bytes. For the RX buffers, you also need to ensure that no other variables are placed just after in the same cache line, as it will be invalidated at some point (not sure how to do that though). I don't think those requirement apply for the TX buffer, but I am not 100% sure.

FRASTM commented:
The cache invalidate operation can be done in async_evt_rx_rdy() to avoid doing it at 2 different places

--> OK, moving to the `async_evt_rx_rdy()`

FRASTM committed:
tests: drivers: uart stm32 run uart_async api on f767 nucleo

This sets the dts of dma for using the uart 6 asynch api.
The stm32f767 has a dma V1 with request 5 for Tx/Rx usart6
The Tx&Rx pins (PG14, PG9) of the usart6 are connected
on the nucleo_f767zi board to pass the test

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
include: linker: define a USER section in the NOCACHE_MEMORY

With this section the user appliaction has a NOCACHE memory
for its own use.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: uart async with DMA w/o cache memory

In this test case, the DMA buffers are allowed in the
USERSPACE NONCACHE MEMORY. This is especially useful
when testing the stm32f7 or h7 soc family.
Because the NOCACHE is NA for the user a new area is created
with RW access.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase
Because the tests/drivers/uart/uart_async_api is defining the USERSPACE, the Tx and Rx buffers are allocated in the USER NOCACHE_MEMORY area which is newly defined by the PR #37917

FRASTM commented:
Re-opening as the discussion is on-going (also with https://github.com/zephyrproject-rtos/zephyr/issues/40146)

For stm32 devices with Cache like stm32F7 when DMA is enabled, a special noncache memory section in the SRAM2 area has been created for DMA buffers.
For devices like the stm32H7, the SRAM2 is well used or in/out DMA buffer

FRASTM commented:
The discussion is on-going with https://github.com/zephyrproject-rtos/zephyr/issues/40146) and https://github.com/zephyrproject-rtos/zephyr/issues/36471
